keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Automated Storage Version Migration with Storage Version Hash
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-01-23
+last-updated: 2019-01-23
+status: provisional
+---
+
+## Table of Contents
+
+* [Goal](#goal)
+* [API design](#api-design)
+* [Storage migration triggering controller](#storage-migration-triggering-controller)
+* [Implications to cluster operators](#implications-to-cluster-operators)
+* [Future work: HA clusters](#future-work-ha-clusters)
+
+## Goal
+
+As the discovery document now exposes the [storage version hash][], storage version
+migration should start automatically when the hash changes.
+
+[storage version hash]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md
+
+## API design
+
+We introduce the `StorageState` API, as a CRD defined in the `migration.k8s.io`
+group.
+
+```golang
+// StorageState is the state of the storage, for a specific resource.
+type StorageState struct {
+  metav1.TypeMeta
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StoageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of versions that the persisted objects might be encoded in.
+  // If empty, it means the stored versions are unknown.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we make it part of the status because
+* Its value can be determined by inspecting the `StorageVersionMigration` API.
```

Comment:
I don't see it here, is this out of date? https://github.com/kubernetes/enhancements/blob/c7ab5663d31b6e3ce848d3929f33a9d6df2eebd1/keps/sig-api-machinery/0030-storage-migration.md

It feels like this should go in spec since it's a historical record and no longer observable from its original source.

Also perhaps we can change the name to PreviousStoredVersionHashes or DisusedStoredVersionHashes or something?

Reply:
Rewrote and added a few points. PTAL.

> Also perhaps we can change the name to PreviousStoredVersionHashes or DisusedStoredVersionHashes or something?

The current storage version is also included in the list, so "Previous..." isn't accurate.

I think "StoredVersionHashes" is intuitive. It means the versions that are possibly persisted in the etcd right now.